### PR TITLE
refactoring to cats effect

### PR DESCRIPTION
**Pull request by katlasik**
- Refactoring of CommitMarkerOffsetsActor
- Refactoring of MarkersQueue

### CODE REVIEW

File: core/src/main/resources/logback.xml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<configuration>
```
Comment (adamw) [inline core/src/main/resources/logback.xml:1]: main sources shouldn't have logback configuration - that should be left to the apps

File: build.sbt [hunk 2]
```diff
@@ -25,7 +28,8 @@ lazy val commonSettings = commonSmlBuildSettings ++ ossPublishSettings ++ Seq(
   versionScheme := Some("semver-spec"),
   scalacOptions ++= Seq("-unchecked", "-deprecation"),
   evictionErrorLevel := Level.Info,
-  ideSkipProject := (scalaVersion.value != scala2_13),
+  addCompilerPlugin("com.olegpy" %% "better-monadic-for" % "0.3.1"),
+  //ideSkipProject := (scalaVersion.value != scala2_13),
```
Comment (adamw) [inline build.sbt:32]: it's useful to keep this setting, as in multi-version scala builds using projectmatrix intellij gets lost

File: core/src/main/scala/com.softwaremill.kmq/redelivery/CommitMarkerOffsetsActor.scala [hunk 3]
```diff
@@ -1,56 +1,75 @@
...
 
-class CommitMarkerOffsetsActor(markerTopic: String, markerOffsetGroupId: String, clients: KafkaClients) extends Actor with StrictLogging {
+import org.typelevel.log4cats.slf4j.Slf4jLogger
 
-  private val consumer = clients.createConsumer(markerOffsetGroupId, classOf[ByteArrayDeserializer], classOf[ByteArrayDeserializer])
-
-  private var toCommit: Map[Partition, Offset] = Map()
-
-  import context.dispatcher
+class CommitMarkerOffsetsActor(
```
Comment (adamw) [inline core/src/main/scala/com.softwaremill.kmq/redelivery/CommitMarkerOffsetsActor.scala:16]: [minor] maybe this should have a private constructor as the only way to create is via `create`

File: core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala [hunk 4]
```diff
@@ -2,172 +2,223 @@ package com.softwaremill.kmq.redelivery
...
-  override def receive: Receive = {
-    case DoConsume =>
-      try {
-        val markers = markerConsumer.poll(OneSecond).asScala
-        val now = System.currentTimeMillis()
-
-        markers.groupBy(_.partition()).foreach { case (partition, records) =>
-          assignedPartitions.get(partition) match {
-            case None =>
-              throw new IllegalStateException(s"Got marker for partition $partition: ${records.map(_.key())}, but partition is not assigned!")
```
Comment (adamw) [inline core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala:107]: I wonder if this will work correctly in case of a rebalance - it's possible that we get some messages, rebalance happens, we remove the partitions from `assignedPartitions`, and only then handle the messages - which will end with an error

maybe enqueueing the partition added/revoked as actor messages would help? (we'd also be able to get rid of the `Ref`) But then, we can receive messages for which we haven't yet added partitions ...

Comment (adamw) [reply core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala:107]: maybe a solution would be to handle parition-revoked as an actor message (which would be handled only *after* the current receive finishes), and instead of handling partition-assigned here, moved it to the consuming logic - when we encounter a marker for an unknown partition, simply create an appropriate consumer (that is - we'd create `AssignedPartition` lazily)

File: core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala [hunk 5]
```diff
@@ -2,172 +2,223 @@ package com.softwaremill.kmq.redelivery
...
+              case Some(ap) =>
+                IO(ap.handleRecords(records, now)) >>
+                  ap.markersQueue.smallestMarkerOffset().traverse { offset =>
+                    commitMarkerOffsetsActor.tell(CommitOffset(partition, offset))
+                  }
+            }
           }
-        }
 
-        assignedPartitions.values.foreach(_.sendRedeliverMarkers(now))
```
Comment (adamw) [inline core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala:117]: `sendRedeliverMarkers` could return an `IO`

File: core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala [hunk 6]
```diff
@@ -2,172 +2,223 @@ package com.softwaremill.kmq.redelivery
...
-      } finally self ! DoConsume
-  }
-
-  private case class AssignedPartition(
-    markersQueue: MarkersQueue, redeliverActor: ActorRef,
-    var latestSeenMarkerTimestamp: Option[Timestamp], var latestMarkerSeenAt: Option[Timestamp]) {
+        }
+        _ <- assignedPartitions.get.flatTap(partitions => IO(partitions.values.foreach(_.sendRedeliverMarkers(now))))
```
Comment (adamw) [inline core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala:198]: `sendRedeliverMarkers` could return an IO

File: core/src/main/scala/com.softwaremill.kmq/redelivery/RedeliverActor.scala [hunk 7]
```diff
@@ -1,45 +1,55 @@
...
+        IO.pure((toRedeliver ++ m, ()))
+      case (toRedeliver, DoRedeliver) =>
+        val hadRedeliveries = toRedeliver.nonEmpty
+        IO(redeliverer.redeliver(toRedeliver))
+          .as((Nil, ()))
+          .guarantee {
+            if (hadRedeliveries)
+              mailbox.offer(DoRedeliver)
+            else
+              mailbox.offer(DoRedeliver).delayBy(1.second)
```
Comment (adamw) [inline core/src/main/scala/com.softwaremill.kmq/redelivery/RedeliverActor.scala:47]: Isn't there a `.start` missing at the end? otherwise the `guarantee` will take up to a second, waiting for this effect to complete - we won't handle other `RedelvierMakrers` messages meanwhile

File: core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala [hunk 8]
```diff
@@ -2,172 +2,223 @@ package com.softwaremill.kmq.redelivery
...
-              }
+  private def stream(
+      assignedPartitions: Ref[IO, Map[Partition, AssignedPartition]],
+      mailbox: Queue[IO, ConsumeMarkersActorMessage],
+      commitMarkerOffsetsActor: CommitMarkerOffsetsActor,
+      markerConsumer: KafkaConsumer[MarkerKey, MarkerValue]
+  ) = Stream
+    .fromQueueUnterminated(mailbox)
+    .evalTap { case DoConsume =>
+      val process = for {
```
Comment (adamw) [inline core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala:176]: maybe extract this to a method? reading this it's hard to follow, so we need to somehow break things down

File: core/src/main/resources/logback.xml (hunk 1 above)
Comment (katlasik) [reply core/src/main/resources/logback.xml]: fixed

File: build.sbt (hunk 2 above)
Comment (katlasik) [reply build.sbt]: fixed

File: core/src/main/scala/com.softwaremill.kmq/redelivery/CommitMarkerOffsetsActor.scala (hunk 3 above)
Comment (katlasik) [reply core/src/main/scala/com.softwaremill.kmq/redelivery/CommitMarkerOffsetsActor.scala]: fixed

File: core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala (hunk 6 above)
Comment (katlasik) [reply core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala]: fixed

File: core/src/main/scala/com.softwaremill.kmq/redelivery/RedeliverActor.scala (hunk 7 above)
Comment (katlasik) [reply core/src/main/scala/com.softwaremill.kmq/redelivery/RedeliverActor.scala]: correct, fixed it

File: core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala (hunk 4 above)
Comment (katlasik) [reply core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala]: What would happen if we just ignore this state (for example just log a warning)?

Comment (adamw) [reply core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala]: hm well couldn't it produce a false warning? my concern here would be that we could get a race condition, where the partition is added (w/ the callback being invoked), and at the same time new messages start arriving. So if this happened, not only we would get a false warning, but we would also loose a message. But  maybe messages aren't delivered until all the callback complete? That would require verification on the kafka side

File: core/src/main/scala/com.softwaremill.kmq/KafkaClientsResourceHelpers.scala [hunk 9]
```diff
@@ -0,0 +1,37 @@
+package com.softwaremill.kmq
+
+import cats.effect.{IO, Resource}
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.apache.kafka.clients.producer.KafkaProducer
+import org.apache.kafka.common.serialization.{Deserializer, Serializer}
+
+import scala.jdk.CollectionConverters._
+
+final class KafkaClientsResourceHelpers(val clients: KafkaClients) {
```
Comment (adamw) [inline core/src/main/scala/com.softwaremill.kmq/KafkaClientsResourceHelpers.scala:10]: this is used only from `redelivery`, let's move it there and make private (as it's a helper)

File: core/src/main/scala/com.softwaremill.kmq/KafkaClientsResourceHelpers.scala [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+      extraConfig: Map[String, AnyRef] = Map.empty
+  ): Resource[IO, KafkaConsumer[K, V]] = Resource.make(
+    IO(clients.createConsumer(groupId, keyDeserializer, valueDeserializer, extraConfig.asJava))
+  )(p => IO(p.close()))
+
+}
+
+object KafkaClientsResourceHelpers {
+
+  def apply(bootstrapServers: String, extraGlobalConfig: Map[String, AnyRef]) =
```
Comment (adamw) [inline core/src/main/scala/com.softwaremill.kmq/KafkaClientsResourceHelpers.scala:34]: is this ever used?

File: core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala [hunk 11]
```diff
@@ -2,180 +2,220 @@ package com.softwaremill.kmq.redelivery
...
-  private var redeliverActorNameCounter = 1
-
-  private var commitMarkerOffsetsActor: ActorRef = _
-
-  override def preStart(): Unit = {
-    markerConsumer = clients.createConsumer(
-      config.getMarkerConsumerGroupId,
-      classOf[MarkerKey.MarkerKeyDeserializer],
-      classOf[MarkerValue.MarkerValueDeserializer]
+  def create(config: KmqConfig, helpers: KafkaClientsResourceHelpers) = {
```
Comment (adamw) [inline core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala:24]: public methods should have types (there's an intellij inspection for that)

**Review by adamw — APPROVED** [review]

**Comment by adamw** [comment]
Great work, thanks! :)

**Comment by alexenvoy** [comment]
As someone working on integrating your project into our system, I'm wondering if I could ask you about the motivation for this change. Is this due to licensing issues with Akka?

**Comment by adamw** [comment]
Yes, exactly - Akka licensing and it's non-viability for OSS projects
